### PR TITLE
Add [Coderabbit] PR Stats service and tests

### PR DESCRIPTION
**Pull request by aravindputrevu**
- Implemented a new service `CoderabbitStats` to fetch and display pull request statistics from the CodeRabbit API.
- Created a corresponding tester file to validate the service's functionality, including tests for valid repositories, repository not found, and server errors.
- The service returns a badge with the number of PRs and appropriate error messages based on the API response.

This addition enhances the analysis capabilities of the application by integrating CodeRabbit statistics.

GitHub Issue - https://github.com/badges/shields/issues/10748

### CODE REVIEW

**Review by jNullj — CHANGES_REQUESTED** [review]
Thank you for the contribution.
I noticed you did not add any live tests (making real API calls to external service) and only mocked responses.
Please give a quick read to [this section of our docs about tests](https://github.com/badges/shields/blob/master/doc/service-tests.md#2-our-first-test-case), as mentioned its best to include at least one live test so we know our service is still live and up to date with upstream.

> Note that when we call our badge, we are allowing it to communicate with an external service without mocking the response. We write tests which interact with external services, which is unusual practice in unit testing. We do this because one of the purposes of service tests is to notify us if a badge has broken due to an upstream API change. For this reason it is important for at least one test to call the live API without mocking the interaction

Is there a specific reason to avoid using live tests? If no could you add at least one live test, Its common for our project to have all calls to live services in tests unless we test internal functionality like data transformation of API response.

**Comment by aravindputrevu** [comment]
@jNullj I have made relevant changes as per your request, requesting you to see if it fits the guidelines.

File: services/coderabbit/coderabbit-stats.service.js
Comment (chris48s) [inline services/coderabbit/coderabbit-stats.service.js:37]: From

https://github.com/badges/shields/blob/master/CONTRIBUTING.md#badge-guidelines

> The left-hand side of a badge should not advertise. It should be a lowercase noun succinctly describing the meaning of the right-hand side.

So rather than `CodeRabbit | 7 Reviews`, this badge should be either `reviews | 7` or `coderabbit reviews | 7`

File: services/coderabbit/coderabbit-stats.service.js
Comment (chris48s) [inline services/coderabbit/coderabbit-stats.service.js:45]: Can we ditch the custom colours here and stick with the standard palette please.
We can just get rid of the `labelColor` in `defaultBadgeData`. All badges should use the default `#555555` for the label (left side of the badge)

For the message (right side), lets go with `blue`. We use this for informational badges in our palette, like this one where all badges get the same colour regardless of the number.

File: services/coderabbit/coderabbit-stats.service.js
Comment (chris48s) [inline services/coderabbit/coderabbit-stats.service.js:12]: Can we make the "noun" part of the route `/prs` or `/pull-requests` here. `/stats` doesn't really describe the information on the badge. It also leaves things more open if we ever wanted to add more CodeRabbit badges.
Docs on badge URLs: https://github.com/badges/shields/blob/master/doc/badge-urls.md

File: services/coderabbit/coderabbit-stats.service.js
Comment (chris48s) [inline services/coderabbit/coderabbit-stats.service.js:15]: `examples` is not how we define the docs for the frontend. LLM slop?

The tutorial covers how to add docs to the frontend by writing an OpenAPI spec for the service endpoint here: https://github.com/badges/shields/blob/master/doc/TUTORIAL.md#44-adding-documentation-to-the-frontend

File: services/coderabbit/coderabbit-stats.service.js
Comment (chris48s) [inline services/coderabbit/coderabbit-stats.service.js:65]: Can we remove this try/catch and `renderError()` function

This will catch all the standard exceptions (like `NotFound` if we get a 404 or `InvalidResponse` if the validation fails, etc) and replace them.

File: services/coderabbit/coderabbit-stats.service.js
Comment (chris48s) [inline services/coderabbit/coderabbit-stats.service.js:54]: Any particular reason not to stick with the default 404 message ("not found") here? If we want to override it, something more descriptive (e.g: "project not found", "repo not found" etc) would be more helpful than "invalid"

File: services/coderabbit/coderabbit-stats.service.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js:12]: @chris48s this would be a major change on the endpoint side as we need to change the backend code. We are on a freeze now and can make this change in Jan 2025.

File: services/coderabbit/coderabbit-stats.service.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js]: I have updated the text accordingly.

File: services/coderabbit/coderabbit-stats.service.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js]: I have removed colors fully

File: services/coderabbit/coderabbit-stats.service.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js]: Sorry, this is a oversight on my end. I was trying to understand the tutorial and also used GH Copilot to get stub code. 

Should be more careful! Apologies, I have made the correction.

File: services/coderabbit/coderabbit-stats.service.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js]: Removed it.

File: services/coderabbit/coderabbit-stats.service.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js]: I have made the error, repo not found. Updated tests.

**Comment by aravindputrevu** [comment]
@chris48s I'm contributing a badge for the first time to shields.io. 

Sorry for the trouble with obvious mistakes. I've read the contributor.md and tutorial.md, but I still missed a lot. Please re-review and let me know if things are out of place.

File: services/coderabbit/coderabbit-stats.service.js
Comment (chris48s) [reply services/coderabbit/coderabbit-stats.service.js:12]: What we are declaring here is the URL for the badge on shields.io

The route on the upstream API (i.e: https://api.coderabbit.ai/stats/... ) is independent of this.

File: services/coderabbit/coderabbit-stats.service.js
Comment (chris48s) [reply services/coderabbit/coderabbit-stats.service.js]: The label text should be lowercase, even where we're including a service name

File: services/coderabbit/coderabbit-stats.tester.js
Comment (jNullj) [inline services/coderabbit/coderabbit-stats.tester.js:24]: Could we use live examples for all types of responses?

**Review by jNullj — CHANGES_REQUESTED** [review]

File: services/coderabbit/coderabbit-stats.tester.js
Comment (jNullj) [reply services/coderabbit/coderabbit-stats.tester.js]: dont forget to import Joi using `import Joi from 'joi'`

File: services/coderabbit/coderabbit-stats.service.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js]: Made this change.

File: services/coderabbit/coderabbit-stats.service.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js]: Sorry for mis-understanding. Made the change

File: services/coderabbit/coderabbit-stats.service.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js]: Remove the "By default" language

File: services/coderabbit/coderabbit-stats.tester.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.tester.js]: Made all tests use the live service endpoint.

File: services/coderabbit/coderabbit-stats.service.js
Comment (chris48s) [inline services/coderabbit/coderabbit-stats.service.js:49]:
```suggestion
    return {
      message: `${reviews}`,
      color: 'blue',
    }
```

File: services/coderabbit/coderabbit-stats.service.js
Comment (chris48s) [inline services/coderabbit/coderabbit-stats.service.js:25]: The example value should be a value that works, so "github" would be sensible value for `example` here. Basically if I copy and paste it into the builder, it should work.

We also don't need to double up examples in the description.

In the docs, if we want to communicate to the user that there is a closed set of valid values, we can give the user a dropdown to pick from. So if you look at something like the `vcsType` param on https://shields.io/badges/coveralls you get a dropdown for [github, bitbucket, gitlab].

Screenshot at 2024-12-23 18-12-35

Here's what that looks like in the code.

https://github.com/badges/shields/blob/0cbc1319ce30e1af262c88c12fb7e0deb890e6f3/services/coveralls/coveralls.service.js#L17

https://github.com/badges/shields/blob/0cbc1319ce30e1af262c88c12fb7e0deb890e6f3/services/coveralls/coveralls.service.js#L26-L30

Lets do the same here.

File: services/coderabbit/coderabbit-stats.service.js
Comment (chris48s) [inline services/coderabbit/coderabbit-stats.service.js:58]: You've set a custom value for 404 here, but in the tests you're asserting the the value on the badge is "invalid" rather than "repo not found". The reason for this is that the upstream API actually returns a `400 Bad Request` when the provider, org or repo is not found for the test cases where the org or repo does not exist.

e.g:

```
$ curl "https://api.coderabbit.ai/stats/github/coderabbitai/invalid-repo-name" --include
HTTP/2 400 
...

{"error":"Failed to get CodeRabbit reviews"}
```

It looks like what you really want here is 

```js
httpErrors: {
   400: 'provider or repo not found',
}
```

I can't find any docs on this API. Do you know if there is any case where it actually does return a 404?

**Review by chris48s — COMMENTED** [review]
Latest draft is looking pretty good. There's a few more bits on this latest revision, but we're getting there. Thanks

File: services/coderabbit/coderabbit-stats.service.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js:59]: @chris48s This is a super valid point; I'd ask the team to push the document to our API reference. 

Will update the message to reflect the 400.

File: services/coderabbit/coderabbit-stats.service.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js:25]: I'm ok for the time-being it showing only GH. As GH has most usage, I'll be happy to update in the next revision with the upcoming badges. 

Editing the text to only show GitHub.

Comment (chris48s) [reply services/coderabbit/coderabbit-stats.service.js]: OK. Are `gitlab` and `bitbucket` values that work today, or not?

If they are, lets make this

```js
static route = {
  base: 'coderabbit',
  pattern: 'prs/:provider(github|bitbucket|gitlab)/:org/:repo',
}
```

in the route and

```js
{
  name: 'provider',
  example: 'github',
  description: 'Version Control Provider',
  schema: { type: 'string', enum: this.getEnum('provider') },
}
```

in the Open API spec.

That will show it as

Screenshot at 2024-12-23 20-21-57

in the docs, and we're good to go.

Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js]: Made this change as you mentioned. Thanks for the suggestion!

File: services/coderabbit/coderabbit-stats.tester.js
Comment (jNullj) [inline services/coderabbit/coderabbit-stats.tester.js:19]: What's the benefit of this test? how does it differ from `live CodeRabbitStats`?
```suggestion

```

File: services/coderabbit/coderabbit-stats.service.js
Comment (jNullj) [inline services/coderabbit/coderabbit-stats.service.js:1]: We started with `CodeRabbitStats` as the first draft used stat for the path.
While iterating over this PR we changed to pull request/prs but left stats at file name and class name.
Should we change filenames and class name fore better consistency? maybe to something like `CodeRabbitPullRequest` and `coderabbit-pull-request` similar to other services.

**Review by jNullj — CHANGES_REQUESTED** [review]
Looks good, i run it and it seems to work as intended.
I have 2 more comments, I think that other then this 2 I could approve the PR.
Take note that we usually wait until all reviewers involved with the PR approve before merging the PR.

File: services/coderabbit/coderabbit-stats.service.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.service.js:1]: Thank you. This is amazing find!

File: services/coderabbit/coderabbit-stats.tester.js
Comment (aravindputrevu) [reply services/coderabbit/coderabbit-stats.tester.js]: Updated the test file.

**Review by jNullj — APPROVED** [review]
LGTM

**Review by chris48s — APPROVED** [review]

**Pull request merged**
